src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AsyncWindowsFileStreamStrategy.cs

```diff
@@ -152,92 +148,60 @@ private unsafe ValueTask<int> ReadAsyncInternal(Memory<byte> destination, Cancel
-                    }
+                    destination = positionBefore <= len ?
+                        destination.Slice(0, (int)(len - positionBefore)) :
+                        default;
                 }
 
                 // Now set the position to read from in the NativeOverlapped struct
                 // For pipes, we should leave the offset fields set to 0.
-                intOverlapped->OffsetLow = unchecked((int)positionBefore);
-                intOverlapped->OffsetHigh = (int)(positionBefore >> 32);
+                nativeOverlapped->OffsetLow = unchecked((int)positionBefore);
+                nativeOverlapped->OffsetHigh = (int)(positionBefore >> 32);
 
                 // When using overlapped IO, the OS is not supposed to
                 // touch the file pointer location at all.  We will adjust it
                 // ourselves, but only in memory. This isn't threadsafe.
                 _filePosition += destination.Length;
             }
 
-            // queue an async ReadFile operation and pass in a packed overlapped
-            int r = FileStreamHelpers.ReadFileNative(_fileHandle, destination.Span, false, intOverlapped, out int errorCode);
-
-            // ReadFile, the OS version, will return 0 on failure.  But
-            // my ReadFileNative wrapper returns -1.  My wrapper will return
-            // the following:
-            // On error, r==-1.
-            // On async requests that are still pending, r==-1 w/ errorCode==ERROR_IO_PENDING
-            // on async requests that completed sequentially, r==0
-            // You will NEVER RELIABLY be able to get the number of bytes
-            // read back from this call when using overlapped structures!  You must
-            // not pass in a non-null lpNumBytesRead to ReadFile when using
-            // overlapped structures!  This is by design NT behavior.
-            if (r == -1)
+            // Queue an async ReadFile operation.
+            bool pending = true;
+            try
             {
-                // For pipes, when they hit EOF, they will come here.
-                if (errorCode == Interop.Errors.ERROR_BROKEN_PIPE)
-                {
-                    // Not an error, but EOF.  AsyncFSCallback will NOT be
-                    // called.  Call the user callback here.
-
-                    // We clear the overlapped status bit for this special case.
-                    // Failure to do so looks like we are freeing a pending overlapped later.
-                    intOverlapped->InternalLow = IntPtr.Zero;
-                    valueTaskSource.ReleaseNativeResource();
-                    TryToReuse(valueTaskSource);
-                    return new ValueTask<int>(0);
-                }
-                else if (errorCode != Interop.Errors.ERROR_IO_PENDING)
+                if (Interop.Kernel32.ReadFile(_fileHandle, (byte*)vts._memoryHandle.Pointer, destination.Length, IntPtr.Zero, nativeOverlapped) == 0)
```

Comment:
Good call. Calling the p/invoke directly simplifies the code by a lot, and we are already doing the error handling directly here.

`ReadFileNative` is still being used by `ConsoleStream`, and the `Net5Compat` and `SyncWindows` strategies. I wonder if we could add a similar improvement on `SyncWindowsFileStreaStrategy` to avoid calling `ReadFileNative`.